Inductor cpp wrapper: fix codegen of ScatterFallback

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104575
* __->__ #104524

Fix cpp wrapper failure on TorchBench model `basic_gnn_edgecnn` and `hf_Reformer` which contain scatter OP.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

chunyuan-w committed:
Inductor cpp wrapper: fix codegen of ScatterFallback

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of ScatterFallback"


Fix cpp wrapper failure on TorchBench model `basic_gnn_edgecnn`:
[src.new_zeros(size).scatter_reduce_(dim, index, src, reduce=f'a{reduce}', include_self=False)](https://github.com/pyg-team/pytorch_geometric/blob/e45059a95cf46f68b10aa890ca497e3937b43811/torch_geometric/utils/scatter.py#L93-L94) where `reduce='amax'`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of ScatterFallback"


Fix cpp wrapper failure on TorchBench model `basic_gnn_edgecnn`:
[src.new_zeros(size).scatter_reduce_(dim, index, src, reduce=f'a{reduce}', include_self=False)](https://github.com/pyg-team/pytorch_geometric/blob/e45059a95cf46f68b10aa890ca497e3937b43811/torch_geometric/utils/scatter.py#L93-L94) where `reduce='amax'`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of ScatterFallback"


Fix cpp wrapper failure on TorchBench model `basic_gnn_edgecnn`:
[src.new_zeros(size).scatter_reduce_(dim, index, src, reduce=f'a{reduce}', include_self=False)](https://github.com/pyg-team/pytorch_geometric/blob/e45059a95cf46f68b10aa890ca497e3937b43811/torch_geometric/utils/scatter.py#L93-L94) where `reduce='amax'`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

jgong5 commented on test/inductor/test_cpp_wrapper.py:
```diff
@@ -179,6 +179,10 @@ class BaseTest(NamedTuple):
         BaseTest("test_relu"),  # multiple inputs
         BaseTest("test_repeat_interleave", "", test_cpu_repro.CPUReproTests()),
         BaseTest("test_scalar_input"),
+        BaseTest("test_scatter4"),
+        BaseTest("test_scatter5"),
+        BaseTest("test_scatter6"),
+        BaseTest("test_scatter_reduce2"),
```
Are we adding all `test_scatter` related tests?

chunyuan-w replied:
Scatter UTs have been added.

jgong5 reviewed (approved):
LGTM. Seems we are not adding all `test_scatter` related UTs in cpp wrapper test? Can we do that?

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of ScatterFallback"


Fix cpp wrapper failure on TorchBench model `basic_gnn_edgecnn` and `hf_Reformer` which contain scatter OP.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of ScatterFallback"


Fix cpp wrapper failure on TorchBench model `basic_gnn_edgecnn` and `hf_Reformer` which contain scatter OP.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge
